docs/configuration.md
---
layout: default
title: Configuration File
has_children: true
nav_order: 5
---
## Configuration File
{: .no_toc }
Since version 2.0, Georgia has included a configuration file named `georgia-config.jsonc`. This file contains various configuration settings for the theme which can be edited to change the appearance and functionality.

## Table of contents
{: .no_toc .text-delta }

* TOC
{:toc}

### Editing

The configuration file format is [JSONC](https://code.visualstudio.com/docs/languages/json#_json-with-comments), JSON with comments. This file can be edited in any text editor, but I recommend VS Code since it handles these files natively. `georgia-config.jsonc` is stored in the root of the `\georgia` folder. To edit the file, you can browse to files location (typically `%appdata%\foobar2000\georgia` if following the standard instructions), or you can open the config file directly from the Options menu:

![Edit config file menu](https://user-images.githubusercontent.com/2282004/108946374-e736e700-7623-11eb-8194-eac839896c58.png)

By default Windows will not know how to handle `.jsonc` files, so you will need to specify a default editor. You can choose VS Code if you have it installed, or select any other text editor you prefer, including notepad.exe.

Because the config file is in the JSONC format, comments are interspersed throughout the file to further explain what the various settings/properties mean. Comments are ignored by Georgia's JSON parser, so you can also ignore or edit them as needed (however if the theme updates the config file, the comments will be added back in).

### Configuration Sections

- `"configVersion"` - Do not edit this unless you are testing certain upgrade scenarios. When the theme loads it will always update this value to the version of Georgia which created the file.
- `"settings"` - Various theme settings. Many of these will be discussed elsewhere in the documentation.
- `"title_format_strings"` - Title formatting strings which the theme uses instead of hardcoding. You can use this to change how certain fields appear, or to get metadata values to show up if you tag your files using different fields than I do (e.g. you can update the value of `title_format_strings.artist_country` if you write these values to `%artist country%` instead of the default value).
- `"metadataGrid"` - This is an array containing all the potential label/value pairs of metadata which will be displayed in the left hand information panel when neither the Playlist or Library views are open. Every entry in the array must contain both a `"label"`, and `"value"` property. Optionally, you can add the `"age": true` property to date metadata values to [append a string showing how long ago that date was from the current date](https://user-images.githubusercontent.com/2282004/108948026-477b5800-7627-11eb-88ee-688f2d18139a.png). If the `value` property evaluates to an empty string, the entire row will be removed from the metadataGrid, meaning you can add rows that only apply to some of the files in your collection and it won't always clutter up the display.
- `"imgPaths"` - An array containing the paths which will be searched to find images to display. If the "Cycle through all artwork" option is disabled, only the first matched image will be shown. If the "Cycle through all artwork" option is enabled, then all images found will rotate through in that order. As configured by default, images matching the `folder.*` pattern will be displayed first, and then the rest of your images will cycle in alphabetical order.
- `"lyricFilenamePatterns"` - An array containing all the possible title formatting strings to match when searching for lyrics files. Do not include paths! These file patterns will be searched for in various locations which are currently not configurable. See the lyrics section for more details.

### Upgrades

Whenever you reload Georgia, the theme will check its hardcoded version with the `configVersion` saved in `georgia-config.jsonc`. If `georgia-config.jsonc` is an older version, the configuration file will be upgraded, and any new or updated properties will be added to the configuration file. This *might* mean that in some rare occasions, a value you have edited might be overwritten with a new default. This will be done sparingly and I will attempt to notify users when it happens. 99% of the time, any existing properties will be left completely untouched.

During upgrades, the old version of the configuration file will be backed up to a new file containing the version of georgia which wrote it. For example, upgrading from 2.0.0 to 2.1.0 will create a `georgia-config-2.0.0.jsonc`. You can use these backups in case something goes wrong and you need to downgrade the version of Georgia, or so that you have a copy of your setting for any properties that might get reset to new defaults during the upgrade.

### Help, I've screwed something up!

So you've edited the config file and now the theme won't load. Just delete the file and reload Georgia and a new one will be written for you. If you have a backup containing changes you've made, just rename the file (e.g. `georgia-config-2.0.0.json` => `georgia-config.jsonc`) and then reload the theme. It will be upgraded for you.

If the theme loads properly but you've borked some of the title format fields you can always reset your config file back to it's default by using the ["Reset config file" menu item](https://user-images.githubusercontent.com/2282004/108946374-e736e700-7623-11eb-8194-eac839896c58.png) in the Options menu.


docs/configuration-titleformatting.md
---
layout: default
title: Title-Formatting Strings
parent: Configuration File
nav_order: 1
---
## Title-Formatting Strings

This following table contains every property in the `title_format_strings` config object, what it's used for, and where it appears in the UI

|Property   |Description   |Location   |
|---|---|---|
|album_subtitle|Usually an un-official name for the album; [Weezer's "Blue Album"](https://user-images.githubusercontent.com/2282004/109090486-9fbe6280-76d8-11eb-99ff-2c091fe2a81b.png)|In brackets and italics next to the album name at the top of the metadata grid and in the playlist header|
|album_translation|A translated title for the album|[Wrapped in brackets next to the album's name](https://user-images.githubusercontent.com/2282004/109090630-e4e29480-76d8-11eb-9fa4-b1975af38658.png) in the metadata grid|
|artist_country|The name or [2-letter abbreviation](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) of the artist's country used for showing flags images. Can be multi-value.|[Beside the artist's name](https://user-images.githubusercontent.com/2282004/109066344-8dc9c900-76b2-11eb-84a8-57c2dc6e5d7f.png)|
|artist|How the artist's name should be formatted|Anywhere the artist is shown|
|date|The full date listed for the album. Shows YYYY-MM-DD if possible.|In the [Playlist header](https://user-images.githubusercontent.com/2282004/109103412-78738f80-76f0-11eb-9c3d-5f7514bcaf9c.png) if ["show full date" is enabled](https://user-images.githubusercontent.com/2282004/109103324-4bbf7800-76f0-11eb-85e6-7a45a424e298.png)|
|disc_subtitle|The subtitle for the current disc, or group of tracks if you wish to separate groups of tracks in the playlist|In the [metadata grid](https://user-images.githubusercontent.com/2282004/109103756-051e4d80-76f1-11eb-90b3-30b4ba467b69.png), and [in the playlist](https://user-images.githubusercontent.com/2282004/109104549-cccb3f00-76f1-11eb-9702-04a67ea14328.png)|
|disc|The disc being played -- by default only shows if more than one disc exists|[Above the progress bar](https://user-images.githubusercontent.com/2282004/109104768-38ada780-76f2-11eb-9edb-c8aebf09ab41.png) and in the playlist [disc row headers](https://user-images.githubusercontent.com/2282004/109104549-cccb3f00-76f1-11eb-9702-04a67ea14328.png)|
|edition|The edition of the album. Typically only used if different than the standard release.|Shown in the playlist [group header](https://user-images.githubusercontent.com/2282004/109104909-84605100-76f2-11eb-9115-5ae27ad60cce.png)|
|last_played|The last time the current song was played|Used to calculate the [right edge of the played section](https://user-images.githubusercontent.com/2282004/109105407-79f28700-76f3-11eb-95be-a744b037d903.png) of the timeline|
|lyrics|Used for the embedded lyrics check|If evaluates to text, will be shown when the lyrics button is pressed|
|original_artist|The original artist that either wrote or performed the song|Shown in the italics next to the title [above the progress bar](https://user-images.githubusercontent.com/2282004/109105622-ebcad080-76f3-11eb-9314-faa5ff319aef.png) and [in the playlist](https://user-images.githubusercontent.com/2282004/109105765-2e8ca880-76f4-11eb-9467-b2321ae24877.png).|
|releaseCountry|The country where this edition was released. Can be the full country name or a [2-letter country code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2).|Only used to display the flag image in the [metadata grid](https://user-images.githubusercontent.com/2282004/109106157-dc985280-76f4-11eb-9adf-c3d5828e4c1c.png)
|title|The title of the currently playing track. By default includes song title translation|[Above the progress bar](https://user-images.githubusercontent.com/2282004/109106427-5f211200-76f5-11eb-93b0-560f42de01fd.png), and anywhere else the song title is shown, but not in the library or playlist|
|tracknum|How the tracknumber should be shown|Above the progress bar and [to the left of the song title](https://user-images.githubusercontent.com/2282004/109106427-5f211200-76f5-11eb-93b0-560f42de01fd.png)|
|vinyl_side|The vinyl side if it exists. "A", "B", etc. Analogous to %discnumber% for non-vinyl albums, but will always be shown if it exists and "Use vinyl style numbering" is enabled|[Above the progress bar](https://user-images.githubusercontent.com/2282004/109106864-49601c80-76f6-11eb-8777-6ad1f384bc03.png) and [in the playlist](https://user-images.githubusercontent.com/2282004/109106924-6ac10880-76f6-11eb-94bd-f4b2841d3cee.png)|
|vinyl_tracknum|The track number on the vinyl side|[Above the progress bar](https://user-images.githubusercontent.com/2282004/109106864-49601c80-76f6-11eb-8777-6ad1f384bc03.png) and [in the playlist](https://user-images.githubusercontent.com/2282004/109106924-6ac10880-76f6-11eb-94bd-f4b2841d3cee.png)|
|year|Only the four-digit year, regardless if %date% contains a YYYY-MM-DD value|In the [upper right corner](https://user-images.githubusercontent.com/2282004/109107085-b83d7580-76f6-11eb-8490-8f0fa317d24d.png) when playing a song and [in the playlist header](https://user-images.githubusercontent.com/2282004/109107150-d4d9ad80-76f6-11eb-8616-7450d9395a9f.png) if ["show full date" is disabled](https://user-images.githubusercontent.com/2282004/109103324-4bbf7800-76f0-11eb-85e6-7a45a424e298.png)|

changelog.md
### v2.0.0b4 - 2021-02-21
Rolls up all changes from the betas below plus
 - Lyrics filename patterns can be specified in the config file
 - Config file can be edited/reset from the settings menu
 - Artwork is no longer reloaded/parsed unnecessarily when changing tracks quickly (prevents flashes of wrong theme color)
 - Added new icon set (and menu option to select) based on icons created by @Zephyr0ck

### v2.0.0b3 - 2021-01-30
 - Now works with foo_ui_hacks to show min/max/close buttons when applicable and moves UI elements accordingly
 - Fixed regression with Queue'd items not showing in playlist
 - Transport button spacing now configurable (thanks @notsigma)
 - Allow filtering out of cd.jpgs from showing with rest of artwork
 - Improved config version upgrades
 - Added option for showing full date in playlist header
 - Lots of code cleanup

### v2.0.0b2 - 2021-01-15
 - Updating config files from previous versions more robust
 - Adding some new properties/settings
 - Showing release country flag if the tf.releaseCountry field is set
 - Replicated theme background on theme startup so on_paint never shows white
 - Stopped text now shows "foobar" and version. This is configurable and you can set it back to "foobar plays music" which was the old default

### v2.0.0b1 - 2021-01-06
 - foo_jscript_panel replaced with foo_spider_monkey_panel
 - Simplified script initialization (no more pasting contents of Georgia.txt into Configuration panel after initial setup time)
 - Automatically generating and reading preferences from georgia-config.jsonc
 - When using hyperlinks to search, if current playing song is in results, it will show as playing
 - Added georgia-config.jsonc file to store preferences outside foobar
 - Updating track information in when `on_playback_dynamic_info_track` is called.
 - Improve visibility of progress bar when art primary color is too dark (i.e. close to the background color)
 - Theme update checks happen once a day if enabled

### v1.1.9 - 2020-07-10
 - Fix library panel not showing tracks with foo_jscript_panel 2.4.x
 - Allow specifying a custom cdart filename
 - No longer show "0000" for date
 - Allow override of playlist row_h
 - Fix issues related to font-sizes in playlist header
 - Prevent labels in playlist header from being drawn over group info
 - Handle hyperlinks searching for albums with editions listed

### v1.1.8 - 2020-05-09
 - Random now actually randomizes playlist
 - Fixed volume control issues
 - Improved tooltip handling for buttons
 - Fixed issues with expanded volume bar disappearing and it's appearance in 4k mode
 - Fixed crash when deleting last playlist
 - CD Rotation values were bogus
 - Refactored all menus using new `Menu` helper class, which cut menu code length in half and made adding new options much easier
 - Fixed crash when using weblinks
 - Playlist row and header fonts are scalable through Options >> Playlist settings
 - Option to move transport controls below artwork
 - Visual improvements in 4k mode (ensuring spacing between elements is scaled correctly)
 - Adding Georgia entries to "Help" menu to quickly debug if the theme is installed correctly
 - Added tooltips on hovering over timeline
 - Adjust menu font sizes through options menu
 - Adjust transport button sizes through options menu

### v1.1.7 - 2020-04-11
 - Invert logos when theme primary color is dark (requires foo_jscript_panel v2.3.6)
 - Fixed crash when clicking the hyperlink to upgrade. Sorry!
 - Fixed crash when managing grouping presets
 - Added volume control
 - Album labels in playlist are now hyperlinks
 - Fixed some date timezone issues
 - Improved playlist look when tags don't have a genre

### v1.1.6 - 2019-11-13
 - Fixed startup crashes when creating buttons
 - Drag & Drop issues
 - Simplified date and timezone handling
 - Cleaned up georgia.txt
 - Improved support for foo_youtube

### v1.1.5 - 2019-10-22
 - Fixes for foo_jscript_panel 2.3.x
 - Removed unneeded files
 - Updating fonts

### v1.1.4 - 2019-08-29
 - Add check for updates

### v1.1.3 - 2019-08-28
 - Fixed broken dates
 - Fixed anti-aliasing on elapsed time when playlist is shown

### v1.1.2 - 2019-08-27
 - Playlist should always draw correctly now
 - Dates should never show as "0000"
 - Year now uses $if3(%original release date%,%originaldate%,%date%)
 - ArtCaching was using the wrong values to scale. Corrected
 - Ticks on the timeline should never show overlap the album art

### v1.1.1 - 2019-08-11
 - Crash on startup when display playlist on startup set

### v1.1.0 - 2019-08-10
 - Dark mode (new default)! Switch between the two in the options menu
 - A ton more 4k fixes
 - reiniting playlist when 4k mode switches to avoid scrollbar issues
 - accurate date difference code based on human accepted norms of what a date difference is (i.e. 1 month ago)
 - correctly handling forbidden characters when attempting to find artwork/files
 - better sorting of results when clicking on hyperlinks
 - searching dates by year only
 - Fixed a bunch of issues with Multi-channel display
 - Highlight colors in library/playlist should still allow text to be legible
 - Drastically reduced console spam

### v1.0.1 - 2019-01-23
 - Fix some 4k scaling issues
 - auto load library 10s after startup for better response time
 - fix crash in jscript 2.2.0+
 - variable font sizing for artist string

### v1.0.0 - First official release

js/settings.js
/** @type {*} */
var pref = new PanelProperties(); // preferences
/** @type {*} */
let settings = {};
/** @type {*} */
let globals = {};
/** @type {*} */
let transport = {};


const currentVersion = '2.0.2';
let configVersion = currentVersion;	// will be overwritten when loaded from config file
let updateAvailable = false;
let updateHyperlink;

const g_component_playcount = utils.CheckComponent('foo_playcount');
const g_component_utils = utils.CheckComponent('foo_utils');
const componentEnhancedPlaycount = utils.CheckComponent('foo_enhanced_playcount');
const componentUiHacks = utils.CheckComponent('foo_ui_hacks');

/** @type {*} */
const doc = new ActiveXObject('htmlfile');
const app = new ActiveXObject('Shell.Application');
/** @type {*} */
const WshShell = new ActiveXObject('WScript.Shell');
/** @type {*} */
const fso = new ActiveXObject('Scripting.FileSystemObject');
/** @type {*} */
const vb = new ActiveXObject('ScriptControl');
/** @type {*} */
const UIHacks = !componentUiHacks || new ActiveXObject('UIHacks');

// THEME PREFERENCES/PROPERTIES EXPLANATIONS - After initial run, these values are changed in Options Menu or by Right Click >> Properties and not here!
pref.add_properties({
	version: ['_theme_version (do not hand edit!)', currentVersion],
	rotation_amt: ['Art: Degrees to rotate CDart', 3], // # of degrees to rotate per track change.
	cycleArt: ['Art: Cycle through all images', true], // true: use glob, false: use albumart reader (front only)
	display_cdart: ['Art: Display CD art', true], // true: show CD artwork behind album artwork. This artwork is expected to be named cd.png and have transparent backgrounds (can be found at fanart.tv)
	art_rotate_delay: ['Art: Seconds to display each art', 30], // seconds per image
	rotate_cdart: ['Art: Rotate CD art on new track', true], // true: rotate cdArt based on track number. i.e. rotationAmt = %tracknum% * x degrees
	cdart_ontop: ['Art: Show CD art above front cover', false], // true: display cdArt above front cover
	spinCdart: ['Art: Spin CD art', false], // true: cdArt will spin while the song plays
	filterCdJpgsFromAlbumArt: ['Art: Filter out cd/vinyl .jpgs from showing as artwork', false],
	labelArtOnBg: ['Art: Draw label art on background', false], // true: don't show the theme color background behind label art
	show_flags: ['Show country flags', true], // true: show the artist country flags
	// check_multich:		['Check for MultiChannel version', false],	// true: search paths in tf.MultiCh_paths to see if there is a multichannel version of the current album available
	use_vinyl_nums: ['Use vinyl style numbering (e.g. A1)', true], // true: if the tags specified in tf.vinyl_side and tf.vinyl_tracknum are set, then we'll show vinyl style track numbers (i.e. "B2." instead of "04.")
	startPlaylist: ['Display playlist on startup', false], // true: show the playlist window when the theme starts up
	show_progress_bar: ['Show Progress Bar', true], // true: show progress bar, otherwise hide it (useful is using another panel for this)
	transport_buttons_size: ['Transport: Button size', 32], // size in pixels of the buttons
	transport_buttons_spacing: ['Transport: Button spacing', 5], // size in pixels of the spacing between buttons

	show_timeline_tooltips: ['Show timeline tooltips', true], // true: show tooltips when hovering over the timeline that show information on plays

	menu_font_size: ['Menu font size', 12],

	freq_update: ['Frequent progress bar updates', true], // true: update progress bar multiple times a second. Smoother, but uses more CPU
    darkMode: ['Use Dark Theme', true], // true: use a darker background
	use_4k: ['Detect 4k', 'auto'], // auto: switch to 4k mode when window width wide enough, never: never use 4k mode, always: always use 4k mode
	checkForUpdates: ['Check for Updates', true], // true: check github repo to determine if updates exist
	loadAsync: ['Load Theme Asynchronously', true],	// loads individual theme files asynchronously at startup to reduce risk of FSM throwing slow script error on startup

	lyrics_normal_color: ['Lyrics: Text Color', 'RGBA(255, 255, 255, 255);'],
	lyrics_focus_color: ['Lyrics: Text Highlite Color', 'RGBA(255, 241, 150, 255);'],

	showPlaylistFulldate: ['Playlist: Show full date', false],
	hyperlinks_ctrl: ['Playlist: Hyperlinks require CTRL Key', false], // true: clicking on hyperlinks only works if CTRL key is held down
	show_weblinks: ['Playlist: Show weblinks', true],

	font_size_playlist: ['Font Size: Playlist', 12],
	font_size_playlist_header: ['Font Size: Playlist Header', 15],
	lyrics_font_size: ['Font Size: Lyrics', 20],
});

// Lyrics variables
// lyrics color definitions
var g_txt_normalcolour = eval(pref.lyrics_normal_color);
var g_txt_highlightcolour = eval(pref.lyrics_focus_color);
var g_txt_shadowcolor = RGBA(0, 0, 0, 255);

// TEXT FIELDS
var stoppedTime = 'Georgia v' + currentVersion;

/* My old ridiculous artist string:
$puts(AF,
	$ifgreater($meta_num(ArtistFilter),1,$puts(mArtist,$meta(ArtistFilter,0))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)\
	$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$get(mArtist)\
	$if($stricmp($get(mArtist),%artist%),$puts(feat,1),)\
	$puts(mArtist,$meta(ArtistFilter,1))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)\
	$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$if($get(feat), feat. ,', ')$get(mArtist)\
	$puts(mArtist,$meta(ArtistFilter,2))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)\
	$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$ifequal($meta_num(ArtistFilter),3,' & ',', ')$get(mArtist)\
	$puts(mArtist,$meta(ArtistFilter,3))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)\
	$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$ifequal($meta_num(ArtistFilter),4,' & ',', ')$get(mArtist)\
	$puts(mArtist,$meta(ArtistFilter,4))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)\
	$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$ifequal($meta_num(ArtistFilter),5,' & ',', ')$get(mArtist)\
	)))))))))),%artist%)
)
$ifequal($strcmp(%album artist%,%artist%),1,$get(AF),$if3($meta(artist),%composer%,%performer%,%album artist%))

In one line for adding to config file:
$puts(AF,$ifgreater($meta_num(ArtistFilter),1,$puts(mArtist,$meta(ArtistFilter,0))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$get(mArtist)$if($stricmp($get(mArtist),%artist%),$puts(feat,1),)$puts(mArtist,$meta(ArtistFilter,1))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$if($get(feat), feat. ,', ')$get(mArtist)$puts(mArtist,$meta(ArtistFilter,2))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$ifequal($meta_num(ArtistFilter),3,' & ',', ')$get(mArtist)$puts(mArtist,$meta(ArtistFilter,3))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$ifequal($meta_num(ArtistFilter),4,' & ',', ')$get(mArtist)$puts(mArtist,$meta(ArtistFilter,4))$if($put(comma,$sub($strstr($get(mArtist),', '),1)),$puts(mArtist,$substr($get(mArtist),$add($get(comma),3),$len($get(mArtist))) $substr($get(mArtist),0,$get(comma))),)$if($get(mArtist),$if($or($stricmp($get(mArtist),'Soundtrack'),$stricmp($get(mArtist),'Various Artists')),,$ifequal($meta_num(ArtistFilter),5,' & ',', ')$get(mArtist))))))))))),%artist%))$ifequal($strcmp(%album artist%,%artist%),1,$get(AF),$if3($meta(artist),%composer%,%performer%,%album artist%))
*/


const configPath = fb.ProfilePath + 'georgia\\georgia-config.jsonc';
const config = new Configuration(configPath);
let titleformat = {};
if (!config.fileExists) {
	settings = config.addConfigurationObject(settingsSchema, settingsDefaults, settingsComments);
	tf = config.addConfigurationObject(titleFormatSchema, defaultTitleFormatStrings, titleFormatComments);
	config.addConfigurationObject(gridSchema, defaultMetadataGrid);	// we don't assign an object here because these aren't key/value pairs and thus can't use the get/setters
	config.addConfigurationObject(imgPathSchema, imgPathDefaults);
	config.addConfigurationObject(lyricFilenamesSchema, lyricFilenamesDefaults);
	config.addConfigurationObject(transportSchema, transportDefaults);
	console.log('> Writing', configPath);
	config.writeConfiguration();
}
if (config.fileExists) {
	const prefs = config.readConfiguration();
	/**
	 * While we've read all the values in, we still need to call addConfigurationObject to add the getters/setters
	 * for the objects so that the file gets automatically written when a setting is changed.
	 **/
	settings = config.addConfigurationObject(settingsSchema, Object.assign({}, settingsDefaults, prefs.settings), settingsComments);
	transport = config.addConfigurationObject(transportSchema, Object.assign({}, transportDefaults, prefs.transport), transportComments);
	tf = config.addConfigurationObject(titleFormatSchema, Object.assign({}, defaultTitleFormatStrings, prefs.title_format_strings), titleFormatComments);
	prefs.metadataGrid.forEach(entry => {
		// copy comments over to existing object so they aren't lost
		const gridEntryDefinition = defaultMetadataGrid.find(gridDefItem => gridDefItem.label === entry.label);
		if (gridEntryDefinition && gridEntryDefinition.comment) {
			entry.comment = gridEntryDefinition.comment;
		}
	});
	config.addConfigurationObject(gridSchema, prefs.metadataGrid);	// can't Object.assign here to add new fields. Add new fields in the upgrade section of migrateCheck
	config.addConfigurationObject(imgPathSchema, prefs.imgPaths);
	config.addConfigurationObject(lyricFilenamesSchema, prefs.lyricFilenamePatterns || lyricFilenamesDefaults);

	/* Safety checks. Fix up potentially bad vals from config */
	settings.cdArtBasename = settings.cdArtBasename && settings.cdArtBasename.trim().length ? settings.cdArtBasename.trim() : 'cd';
	settings.artworkDisplayTime = Math.min(Math.max(settings.artworkDisplayTime, 5),120);	// ensure min of 5sec and max of 120sec

	globals.imgPaths = prefs.imgPaths;
	globals.lyricFilenamePatterns = prefs.lyricFilenamePatterns;
	metadataGrid = prefs.metadataGrid;
	configVersion = prefs.configVersion || prefs.version;
	// when adding new objects to the config file, add them in the version check below
}

// do the migration check BEFORE we start adding extra crap to tf.
// TODO: Should I move all the tf. extra properties that AREN'T in the config to the globals object? Probably maybe?
migrateCheck(currentVersion, configVersion);

/* All tf values from here below will NOT be written to the georgia-config file */
tf.vinyl_track = '$if2(' + tf.vinyl_side + '[' + tf.vinyl_tracknum + ']. ,[%tracknumber%. ])';

tf.lyr_path = [ // simply add, change or re-order entries as needed
	'$replace($replace(%path%,%filename_ext%,),\,\\)',
	fb.ProfilePath + 'lyrics\\',
	fb.FoobarPath + 'lyrics\\',
];

tf.labels = [ // Array of fields to test for publisher. Add, change or re-order as needed.
	'label', // DO NOT put %s around the field names because we are using $meta() calls
	'publisher',
	'discogs_label'
];

// CD-ART SETTINGS
// we expect cd-art will be in .png with transparent background, best found at fanart.tv.
pref.vinylside_path = '$directory_path(%path%)\\vinyl$if2(' + tf.vinyl_side + ',).png' // vinyl cdart named vinylA.png, vinylB.png, etc.
pref.vinyl_path = '$directory_path(%path%)\\vinyl.png' // vinyl cdart named vinylA.png, vinylB.png, etc.
pref.cdartdisc_path = '$directory_path(%path%)\\' + settings.cdArtBasename + '$ifgreater(%totaldiscs%,1,%discnumber%,).png'; // cdart named cd1.png, cd2.png, etc.
pref.cdart_path = '$directory_path(%path%)\\' + settings.cdArtBasename + '.png'; // cdart named cd.png (or whatever custom value was specified). This is the most common single disc case.
pref.cdart_amount = 0.48; // show 48% of the CD image if it will fit on the screen

function migrateCheck(version, storedVersion) {
	/**
	 * Adds or Replaces value in the grid with updated string from defaults
	 * @param {MetadataGridEntry[]} grid
	 * @param {string} label Label of the value to add or replace
	 * @param {number} position 0-based index of place to insert new value if existing entry not found
	 */
	const replaceGridEntry = (grid, label, position) => {
		const entryIdx = grid.findIndex(gridEntry => gridEntry && gridEntry.label.toLowerCase() === label.toLowerCase());
		const newVal = defaultMetadataGrid[defaultMetadataGrid.findIndex(e => e && e.label.toLowerCase() === label.toLowerCase())];
		if (entryIdx >= 0) {
			grid[entryIdx] = newVal;
		} else {
			grid.splice(position, 0, newVal);
		}
	}

	if (version !== storedVersion) {
		const configFile = config.readConfiguration();
		/** @type {MetadataGridEntry[]} */
		const grid = configFile.metadataGrid;

		// this function clears default values which have changed
		switch (storedVersion) {

			case '2.0.0-beta1':
			case '2.0.0-beta2':
				window.SetProperty('Art: Seconds to display each art', null);
				tf.lyrics = defaultTitleFormatStrings.lyrics;
			case '2.0.0-beta3':
				config.addConfigurationObject(lyricFilenamesSchema, lyricFilenamesDefaults);
			case '2.0.0-beta4':
				settings.extraTrackInfo = settingsDefaults.extraTrackInfo;
				config.addConfigurationObject(transportSchema, transportDefaults);

			case '2.0.0':
			case '2.0.1':
				settings.defaultSortString = settingsDefaults.defaultSortString;
				tf.releaseCountry = defaultTitleFormatStrings.releaseCountry;
				replaceGridEntry(grid, 'Label', 5);
				replaceGridEntry(grid, 'Catalog #', 6);
				replaceGridEntry(grid, 'Release Country', 7);
				config.addConfigurationObject(gridSchema, grid);

				// this block should appear after all previous versions have fallen through
				console.log('> Upgrading Georgia Theme settings from', storedVersion);
				const fileName = `georgia\\georgia-config-${storedVersion}.jsonc`;
				console.log(`> Backing up Georgia Configuration file to ${fileName}`);
				fso.CopyFile(configPath, fb.ProfilePath + fileName);
				config.writeConfiguration();
				window.Reload();


			default:
				break;

		}
	}
	pref.version = currentVersion;	// always update the version panel property
}

let retryCount = 0;	// don't hammer if it's not working

function checkForUpdates(openUrl) {
	var url = 'https://api.github.com/repos/kbuffington/Georgia/tags';
	makeHttpRequest('GET', url, function (resp) {
		try {
			var respObj = JSON.parse(resp);
			updateAvailable = isNewerVersion(currentVersion, respObj[0].name);
			console.log('Current released version of Georgia: v' + respObj[0].name);
			if (updateAvailable) {
				console.log('>>> Georgia update available. Download it here: https://github.com/kbuffington/Georgia/releases');
				updateHyperlink = new Hyperlink('Update Available', ft.lower_bar, 'update', 0, 0, window.Width);
				if (updateHyperlink) {
					stoppedTime += ' - ';
					if (!fb.IsPlaying) {
						str.time = stoppedTime;
						RepaintWindow();
					}
					if (openUrl) {
						updateHyperlink.click();
					}
				}
			} else {
				console.log('You are on the most current version of Georgia');
			}
		} catch (e) {
			if (!updateHyperlink && retryCount < 3) {
				// updateHyperlink failed to be created somehow. Let's check again after 1 minute.
				retryCount++;
				updateAvailable = false;
				scheduleUpdateCheck(61000);
			}
		}
	});
}


/**
 * Schedule an update check. Set at startup and then typically every 24 hours after unless an update is found
 * @param {number} delay in milliseconds
 */
let updateTimer;
function scheduleUpdateCheck(delay) {
	clearTimeout(updateTimer);
	updateTimer = setTimeout(() => {
		if (!updateAvailable) {
			checkForUpdates(false);
			scheduleUpdateCheck(1000 * 60 * 60 * 24);	// check every 24 hours
		}
	}, delay)
}
